File: services/ai/src/components/trainer/trainer.ts
import { PongGame } from "@components/game/game";
import { PongAgent } from "@components/agent/agent";
import { Observer } from "@segfaultx/observable";
import { AgentConfig } from "@core/types/t_game";
import { e_TAG_PLAYER } from "@core/constants/constants";
import { Dqn } from "../dqn/dqn";
import { access, constants, FileHandle, open } from "node:fs/promises"

export class	Trainer {

	private	avg_reward_best = -Infinity;
	private	baseurl_game = "http://localhost:3000/v1/train/";
	private	game_counter = 1;
	private	game: PongGame = new PongGame(`${this.baseurl_game}${this.game_counter}`);
	private	agent: PongAgent;
	private	averager: number[] = new Array(100).fill(0);
	private	sync_frames: number = 2e4;
	private	csv_path = "runs/train_log.csv";
	private	file_handle?: FileHandle;
	private	episodes: number = 0;
	private	listener_episode: Observer<boolean> = new Observer(async (is_done: boolean) => {
		if (! is_done)
			return ;
		this.episodes++;
		this.ft_appendAverager(this.agent.cumulative_reward, this.averager);
		const	average = this.ft_average(this.averager);
		if (average > this.avg_reward_best) {
			this.avg_reward_best = average;
			await this.agent.online_network.save("file://model/agent.model");
			console.log(`Online Network saved`);
		}
		console.log(`Frame_agent: ${this.agent.frame_count}\n
					cumulativeRewardAgent: ${this.agent.cumulative_reward}\n
					cummulativeRewardAverage_two: ${average}\n
					epsilon_two: ${this.agent.epsilon.toFixed(3)}\n
					hits_counter: ${this.agent.hits_counter}\n
					episodes: ${this.episodes}\n
					`);
		let	log = `${Date.now()},`;
			log += `${this.agent.frame_count},`;
			log += `${this.agent.cumulative_reward},`;
			log += `${average},`;
			log += `${this.agent.epsilon},`;
			log += `${this.agent.hits_counter},`;
			log += `${this.episodes}`;
		await this.file_handle?.write(`${log}\n`);
		if (this.episodes % 250 == 0) {
			const	episode_cycle = this.episodes / 250;
			await this.agent.online_network.save(`file://model/checkpoint_ep${episode_cycle}.model`);
			console.log(`Checkpoint ${episode_cycle} saved.`);
		}
		this.agent.ft_resetEpisode();
	});
	private	listener: Observer<boolean> = new Observer(async (is_done: boolean) => {
		if (! is_done)
			return ;
		this.game.url = `${this.baseurl_game}${this.game_counter}`;
		this.game.ft_reset();
		this.agent.ft_reset();
		this.agent.ft_connectGame();
	});

	constructor(config: Omit<AgentConfig, "tag">) {
		this.game.ft_listenWhenDone(this.listener);
		this.agent = new PongAgent(this.game, {tag: e_TAG_PLAYER.TWO, ...config});
		this.agent.ft_listenWhenDoneEpisode(this.listener_episode);
	}

	public	async	ft_build_trainer(): Promise<void> {
		try {
			await access(this.csv_path, constants.W_OK);
		}
		catch (_error) {
			this.file_handle = await open(this.csv_path, "a");
			/*[PENDING][PINNED][URGENT]: Finish with the writing for the columns in the csv.
					* and then implement the append each log when an episode finish in the listener_episode.*/
					/*[PENDING][URGENT]: The agent is not working well. debug it!!!!!!*/
			let	columns = "t_elapsed_s,";
			columns += "frame,";
			columns += "cumulativeRewardAgent,";
			columns += "cumulativeRewardAverage,";
			columns += "epsilon,";
			columns += "hits_counter,";
			columns += "episodes";
			this.file_handle.write(`${columns}\n`);
		}
	}

	public async	ft_train() {
		this.agent.ft_connectGame();
		this.agent.subject_frame_counter.ft_subscribe(new Observer((frame_count) => {
			if (this.agent.ft_isReplayMemoryFilled()) {
//				const	start = Date.now();
//				console.log(`Start: ${start}`);
				if (frame_count % 10 == 0)
					this.agent.ft_trainOnReplayBatch(128, 0.99);
//				const	end = Date.now();
//				console.log(`End: ${end}, diff: ${end - start}`);
				if (frame_count % this.sync_frames == 0) {
					Dqn.ft_copyWeights(this.agent.target_network, this.agent.online_network);
					console.log('Sync\'ed weights from online network to target network agent_two');
				}
			}
		}));
	}

	private	ft_appendAverager(avg: number, averager: number[]): void {
		averager.shift();
		averager.push(avg);
	}

	private	ft_average(averager: number[]): number {
		return (averager.reduce((current, prev) => current + prev) / averager.length);
	}
}

File: services/ai_python/src_py/components/trainer/trainer.py
import os
import csv
import time
from datetime import datetime

from src_py.components.game.game import PongGame
from src_py.components.agent.agent import PongAgent
from src_py.utils.observable import Observer
from src_py.types.t_game import AgentConfig
from src_py.constants.constants import e_TAG_PLAYER
from src_py.components.dqn.dqn import Dqn

class Trainer:
    def __init__(self, config_without_tag: dict, csv_path: str = "runs/train_log.csv"):
        # timing
        self.t0 = time.time()
        self.t_prev_ms = int(self.t0 * 1000)

        # counters & trackers
        self.fc_prev = {"one": 0, "two": 0}
        self.avg_reward_best = float("-inf")
        self.baseurl_game = "ws://localhost:3000/v1/train/"  # tip: ws:// for websocket-client
        self.game_counter = 1
        self.game = PongGame(f"{self.baseurl_game}{self.game_counter}")

        cfg = AgentConfig(tag=e_TAG_PLAYER.TWO, **config_without_tag)
        self.ag_two = PongAgent(self.game, cfg)

        self.averager_two = [0.0 for _ in range(100)]
        self.sync_frames = int(2e4)
        self.episodes = 0

        # CSV setup
        self.csv_path = csv_path
        os.makedirs(os.path.dirname(self.csv_path) or ".", exist_ok=True)
        if not os.path.exists(self.csv_path):
            with open(self.csv_path, "w", newline="") as f:
                w = csv.writer(f)
                w.writerow([
                    "t_wall_iso",           # wall-clock timestamp (ISO 8601)
                    "t_elapsed_s",          # seconds since training start
                    "frame_ag_two",
                    "cumulativeRewardAgent",
                    "cumulativeRewardAverage_two",
                    "epsilon_two",
                    "hits_counter",
                    "episodes"
                ])

        def _listener_episode(is_done: bool):
            if not is_done:
                return
            self.episodes += 1
            # update averages and maybe save best
            self.ft_appendAverager(self.ag_two.cumulative_reward, self.averager_two)
            average_two = self.ft_average(self.averager_two)

            if average_two > self.avg_reward_best:
                self.avg_reward_best = average_two
                os.makedirs("src_py/model", exist_ok=True)
                self.ag_two.online_network.save("src_py/model/agent_two.keras")
                print("Online Network saved")

            # console print (unchanged)
            print(f"""
                Frame_ag_two: {self.ag_two.frame_count}
                cumulativeRewardAgent: {self.ag_two.cumulative_reward}
                cummulativeRewardAverage_two: {average_two}
                epsilon_two: {self.ag_two.epsilon:.3f}
                hits_counter: {self.ag_two.hits_counter}
                episodes: {self.episodes}
                """)

            # === CSV logging (one row per finished episode) ===
            t_now = time.time()
            with open(self.csv_path, "a", newline="") as f:
                w = csv.writer(f)
                w.writerow([
                    datetime.fromtimestamp(t_now).isoformat(timespec="seconds"),
                    round(t_now - self.t0, 3),
                    int(self.ag_two.frame_count),
                    float(self.ag_two.cumulative_reward),
                    float(average_two),
                    float(self.ag_two.epsilon),
                    int(self.ag_two.hits_counter),
                    int(self.episodes)
                ])


            # 💾 Checkpoint every 50 episodes
            if self.episodes % 250 == 0:
                ckpt_path = f"src_py/model/checkpoint_ep{self.episodes / 250}.keras"
                self.ag_two.online_network.save(ckpt_path)
                print(f"Checkpoint saved at {ckpt_path}")
            self.ag_two.ft_resetEpisode()

        def _listener(is_done: bool):
            if not is_done:
                return
            # (optional) switch game instance url as in your TS code
            self.game.url = f"{self.baseurl_game}{self.game_counter}"
            self.game.ft_reset()
            # reset & reconnect
            self.ag_two.ft_reset()
            self.ag_two.ft_connectGame()

        self.listener = Observer[bool](_listener)
        self.listener_episode = Observer[bool](_listener_episode)
        self.game.ft_listenWhenDone(self.listener)
        self.ag_two.ft_listenWhenDoneEpisode(self.listener_episode)

    def ft_train(self):
        self.ag_two.ft_connectGame()

        def _on_frame(frame_count: int):
            if self.ag_two.ft_isReplayMemoryFilled():
                #start = time.time()  # current time in seconds
                #print(f"Start: {start}")
                if frame_count % 10 == 0:
                    self.ag_two.ft_trainOnReplayBatch(128, 0.99)
                #end = time.time()
                #print(f"End: {end}, diff: {end - start}")
                if frame_count % self.sync_frames == 0:
                    Dqn.ft_copyWeights(self.ag_two.target_network, self.ag_two.online_network)
                    print("Sync'ed weights from online network to target network agent_two")

        self.ag_two.subject_frame_counter.ft_subscribe(Observer(_on_frame))

    @staticmethod
    def ft_appendAverager(avg: float, averager: list[float]) -> None:
        averager.pop(0)
        averager.append(avg)

    @staticmethod
    def ft_average(averager: list[float]) -> float:
        return sum(averager) / len(averager) if averager else 0.0
